File: legacy_backend/aggregators/tagesschau.py
"""Aggregator for Tagesschau RSS feeds."""

import copy
import html
import json

from bs4 import BeautifulSoup
from pydantic import BaseModel, Field

from .base import BaseAggregator, RawArticle


class TagesschauAggregatorConfig(BaseModel):
    id: str
    type: str = Field(pattern="^(managed|custom|social)$")
    name: str = Field(min_length=1)
    url: str = ""
    description: str = Field(min_length=1)
    wait_for_selector: str | None = None
    selectors_to_remove: list[str] = Field(default_factory=list)
    options: dict = Field(default_factory=dict)


class TagesschauAggregator(BaseAggregator):
    """Aggregator for Tagesschau.de (German news)."""

    id = "tagesschau"
    type = "managed"
    name = "Tagesschau"
    url = "https://www.tagesschau.de/xml/rss2/"
    description = "Specialized aggregator for Tagesschau.de (German news). Extracts article content using textabsatz paragraphs, embeds video/audio headers when present, and filters out video news and podcasts."
    wait_for_selector = "p.textabsatz"
    selectors_to_remove = [
        "div.teaser",
        "div.socialbuttons",
        "aside",
        "nav",
        "button",
        "div.bigfive",
        "div.metatextline",
        "script",
        "style",
        "iframe",
        "noscript",
        "svg",
    ]

    def __init__(self):
        super().__init__()
        TagesschauAggregatorConfig(
            id=self.id,
            type=self.type,
            name=self.name,
            url=self.url,
            description=self.description,
            wait_for_selector=self.wait_for_selector,
            selectors_to_remove=self.selectors_to_remove,
            options=self.options,
        )

    def should_skip_article(self, article: RawArticle) -> tuple[bool, str | None]:
        """Skip video news and podcasts."""
        # Check title filters
        skip_terms = [
            "tagesschau",
            "tagesthemen",
            "11KM-Podcast",
            "Podcast 15 Minuten",
        ]
        if any(term in article.title for term in skip_terms):
            return True, f"Skipping filtered content: {article.title}"

        # Check URL filters
        if "bilder/blickpunkte" in article.url:
            return True, f"Skipping image gallery: {article.title}"

        # Use base skip logic for age/existence checks
        return super().should_skip_article(article)

    def _extract_media_header(self, soup: BeautifulSoup) -> str | None:
        """
        Extract video or audio header embed code from Tagesschau article page.

        Tagesschau uses a custom player system with:
        - div.v-instance with data-v-type="MediaPlayer"
        - data-v attribute containing JSON with media information
        - embedCode in pluginData.sharing@web.embedCode

        Returns:
            HTML string with media embed iframe (and image if available), or None if no media found
        """
        # Look for Tagesschau media player instances
        # They use: <div class="v-instance" data-v-type="MediaPlayer" data-v="{...}">
        # Look for players in the header/teaser area first (teaser-top class)
        media_players = soup.find_all(
            "div",
            attrs={"data-v-type": "MediaPlayer"},
            class_=lambda x: x and "mediaplayer" in str(x).lower(),
        )

        # Prefer teaser-top players (header media) over other players
        teaser_players = [
            p
            for p in media_players
            if p.get("class")
            and any("teaser-top" in str(c) for c in p.get("class", []))
        ]
        players_to_check = teaser_players if teaser_players else media_players

        for player_div in players_to_check:
            data_v = player_div.get("data-v")
            if not data_v:
                continue

            try:
                # Parse the JSON data (it's HTML-encoded)
                data_v_decoded = html.unescape(data_v)
                player_data = json.loads(data_v_decoded)

                # Check if it's video or audio
                mc = player_data.get("mc", {})
                streams = mc.get("streams", [])

                # Determine if it's audio-only or video
                is_audio_only = all(
                    stream.get("isAudioOnly", False) for stream in streams
                )

                # Extract image/poster from player data
                image_url = None
                # Check common image fields in mc
                for image_field in ["poster", "image", "thumbnail", "preview", "cover"]:
                    if image_field in mc and mc[image_field]:
                        image_url = mc[image_field]
                        break

                # If not found in mc, check streams
                if not image_url:
                    for stream in streams:
                        for image_field in [
                            "poster",
                            "image",
                            "thumbnail",
                            "preview",
                            "cover",
                        ]:
                            if image_field in stream and stream[image_field]:
                                image_url = stream[image_field]
                                break
                        if image_url:
                            break

                # If still not found, check for image elements near the player
                if not image_url:
                    # Look for img tags in parent or sibling elements
                    parent = player_div.find_parent()
                    if parent:
                        # Check for images in the same container
                        img = parent.find("img")
                        if img and img.get("src"):
                            image_url = img.get("src")
                        # Also check for images in previous/next siblings
                        if not image_url:
                            prev_sibling = player_div.find_previous_sibling()
                            if prev_sibling:
                                img = prev_sibling.find("img")
                                if img and img.get("src"):
                                    image_url = img.get("src")

                # Make image URL absolute if found
                if image_url:
                    if image_url.startswith("//"):
                        image_url = "https:" + image_url
                    elif image_url.startswith("/"):
                        image_url = "https://www.tagesschau.de" + image_url
                    self.logger.debug(f"Found player image: {image_url}")

                # Try to extract embed code from pluginData
                plugin_data = player_data.get("pluginData", {})
                sharing_data = plugin_data.get("sharing@web", {})
                embed_code = sharing_data.get("embedCode", "")

                if embed_code:
                    # The embed code is HTML-encoded, decode it
                    embed_code_decoded = html.unescape(embed_code)
                    # Extract iframe src from embed code
                    embed_soup = BeautifulSoup(embed_code_decoded, "html.parser")
                    iframe = embed_soup.find("iframe")
                    if iframe and iframe.get("src"):
                        src = iframe.get("src")
                        # Remove $params$ placeholder if present
                        src = src.replace("$params$", "")
                        # Make sure URL is absolute
                        if src.startswith("//"):
                            src = "https:" + src
                        elif src.startswith("/"):
                            src = "https://www.tagesschau.de" + src
                        # Adjust height for audio (smaller) vs video
                        height = "200" if is_audio_only else "315"
                        # Build media header with iframe
                        # For audio with image: use image as background
                        if is_audio_only and image_url:
                            media_html = (
                                '<div class="media-header" style="position: relative; background-image: url(\''
                                + image_url
                                + "'); background-size: cover; background-position: center; padding: 20px; border-radius: 8px;\">"
                                f'<iframe src="{src}" width="100%" height="{height}" '
                                'frameborder="0" allowfullscreen scrolling="no" style="background: rgba(0,0,0,0.7); border-radius: 4px;"></iframe>'
                                "</div>"
                            )
                        else:
                            # For video, just embed the iframe (it may have its own poster)
                            media_html = (
                                '<div class="media-header">'
                                f'<iframe src="{src}" width="100%" height="{height}" '
                                'frameborder="0" allowfullscreen scrolling="no"></iframe>'
                                "</div>"
                            )
                        return media_html

                # Fallback: construct player from media URL if available
                for stream in streams:
                    media_items = stream.get("media", [])
                    for media_item in media_items:
                        url = media_item.get("url")
                        mime_type = media_item.get("mimeType", "")
                        if not url:
                            continue

                        # Build media header with player (image merged as poster/background)
                        if is_audio_only and "audio" in mime_type.lower():
                            # Create HTML5 audio player with image as background
                            if image_url:
                                media_html = (
                                    '<div class="media-header" style="position: relative; background-image: url(\''
                                    + image_url
                                    + "'); background-size: cover; background-position: center; padding: 20px; border-radius: 8px;\">"
                                    f'<audio controls style="width: 100%; background: rgba(0,0,0,0.7); border-radius: 4px;">'
                                    f'<source src="{url}" type="{mime_type}">'
                                    "Your browser does not support the audio element."
                                    "</audio>"
                                    "</div>"
                                )
                            else:
                                media_html = (
                                    '<div class="media-header">'
                                    f'<audio controls style="width: 100%;">'
                                    f'<source src="{url}" type="{mime_type}">'
                                    "Your browser does not support the audio element."
                                    "</audio>"
                                    "</div>"
                                )
                        elif not is_audio_only and "video" in mime_type.lower():
                            # Create HTML5 video player with poster image
                            poster_attr = f'poster="{image_url}"' if image_url else ""
                            media_html = (
                                '<div class="media-header">'
                                f'<video controls {poster_attr} style="max-width: 100%; height: auto; width: 100%;">'
                                f'<source src="{url}" type="{mime_type}">'
                                "Your browser does not support the video element."
                                "</video>"
                                "</div>"
                            )
                        else:
                            return None

                        return media_html

            except (json.JSONDecodeError, KeyError, AttributeError) as e:
                self.logger.debug(f"Failed to parse Tagesschau media player data: {e}")
                continue

        return None

    def get_header_image_url(self, article: RawArticle) -> str | None:
        """
        Extract media header (video/audio) from Tagesschau article page.

        This method detects Tagesschau media players and stores the embed HTML
        in a custom attribute for use in standardize_format.

        Returns:
            None (media is handled separately in standardize_format)
        """
        try:
            soup = BeautifulSoup(article.html, "html.parser")
            media_header = self._extract_media_header(soup)

            if media_header:
                # Store media header HTML in article object for use in standardize_format
                # We use a custom attribute that won't conflict with Pydantic
                if not hasattr(article, "_media_header_html"):
                    article._media_header_html = media_header
                else:
                    article._media_header_html = media_header
                self.logger.info(
                    f"Found media header (video/audio) for article: {article.title}"
                )
                # Return None to skip image extraction (media takes priority)
                return None
        except Exception as e:
            self.logger.debug(f"Error extracting media header: {e}")

        # Return None to use default image detection
        return None

    def standardize_format(
        self, article: RawArticle, header_image_url: str | None = None
    ) -> None:
        """
        Standardize content format, adding media header (video/audio) if found.

        Overrides base method to prepend Tagesschau media headers before
        the standard header image processing.
        """
        # Check if we have a media header stored from get_header_image_url
        media_header_html = getattr(article, "_media_header_html", None)

        if media_header_html:
            # Prepend media header to content
            article.html = media_header_html + article.html
            self.logger.debug("Prepended media header to article content")
            # Custom handling: use standardize_content_format directly to skip image extraction
            from aggregators.base.process import standardize_content_format

            article.html = standardize_content_format(
                article.html,
                article,
                generate_title_image=False,  # Skip image extraction when media header exists
                add_source_footer=self.feed.add_source_footer if self.feed else True,
                header_image_url=None,
            )
        else:
            # Normal flow: call parent method
            super().standardize_format(article, header_image_url=header_image_url)

    def extract_content(self, article: RawArticle) -> None:
        """Extract content from textabsatz paragraphs."""
        try:
            soup = BeautifulSoup(article.html, "html.parser")
            content = soup.new_tag("div", **{"class": "article-content"})

            # Extract text content
            for element in soup.find_all(["p", "h2"]):
                if element.find_parent(
                    ["div"],
                    class_=lambda x: x
                    and any(
                        skip in x
                        for skip in ["teaser", "bigfive", "accordion", "related"]
                    )
                    if isinstance(x, list)
                    else (
                        x
                        and any(
                            skip in x
                            for skip in ["teaser", "bigfive", "accordion", "related"]
                        )
                    ),
                ):
                    continue
                if element.name == "p" and element.get("class"):
                    classes = element.get("class", [])
                    if any("textabsatz" in c for c in classes):
                        new_p = copy.copy(element)
                        new_p.attrs = {}
                        content.append(new_p)
                elif element.name == "h2":
                    classes = element.get("class", [])
                    if any("trenner" in c for c in classes):
                        new_h2 = soup.new_tag("h2")
                        new_h2.string = element.get_text(strip=True)
                        content.append(new_h2)
            article.html = str(content)
        except Exception as e:
            self.logger.error(
                f"Extraction failed for {article.url}: {e}", exc_info=True
            )


def aggregate(feed, force_refresh=False, options=None):
    aggregator = TagesschauAggregator()
    return aggregator.aggregate(feed, force_refresh, options or {})

File: src/server/aggregators/tagesschau.ts
/**
 * Tagesschau aggregator.
 *
 * Specialized aggregator for Tagesschau.de (German news).
 * Extracts article content using textabsatz paragraphs, embeds video/audio headers when present,
 * and filters out video news and podcasts.
 */

import * as cheerio from "cheerio";

import { logger } from "../utils/logger";

import type { RawArticle } from "./base/types";
import { sanitizeHtml, createHeaderElementFromUrl } from "./base/utils";
import { FullWebsiteAggregator } from "./full_website";

export class TagesschauAggregator extends FullWebsiteAggregator {
  override readonly id = "tagesschau";
  override readonly type: "managed" | "custom" | "social" = "managed";
  override readonly name = "Tagesschau";
  override readonly url = "https://www.tagesschau.de/xml/rss2/";
  override readonly description =
    "Tagesschau.de - German public broadcasting news website providing national and international news coverage.";
  override readonly prefillName = true;
  override readonly identifierEditable = false;

  // Store original HTML for media header extraction
  private originalHtmlCache: Map<string, string> = new Map();

  override readonly waitForSelector = "p.textabsatz";
  override readonly selectorsToRemove = [
    "div.teaser",
    "div.socialbuttons",
    "aside",
    "nav",
    "button",
    "div.bigfive",
    "div.metatextline",
    "script",
    "style",
    "iframe",
    "noscript",
    "svg",
  ];

  protected override shouldSkipArticle(article: RawArticle): boolean {
    // Skip livestream articles
    if (article.title.includes("Livestream:")) {
      logger.info({ title: article.title }, "Skipping livestream article");
      return true;
    }

    // Check title filters
    const skipTerms = [
      "tagesschau",
      "tagesthemen",
      "11KM-Podcast",
      "Podcast 15 Minuten",
      "15 Minuten:",
    ];

    if (skipTerms.some((term) => article.title.includes(term))) {
      logger.info({ title: article.title }, "Skipping filtered content");
      return true;
    }

    // Check URL filters
    if (article.url.includes("bilder/blickpunkte")) {
      logger.info({ title: article.title }, "Skipping image gallery");
      return true;
    }

    return super.shouldSkipArticle(article);
  }

  /**
   * Extract video or audio header embed code from Tagesschau article page.
   */
  private async extractMediaHeader(
    soup: cheerio.CheerioAPI,
  ): Promise<string | null> {
    const playersToCheck = this.getMediaPlayers(soup);

    for (let i = 0; i < playersToCheck.length; i++) {
      const playerDiv = playersToCheck.eq(i);
      const dataV = playerDiv.attr("data-v");
      if (!dataV) continue;

      try {
        const playerData = this.parsePlayerData(dataV);
        const streams = playerData.mc?.streams || [];
        const isAudioOnly = streams.length > 0 && streams.every((s: any) => s.isAudioOnly === true);
        const imageUrl = this.getPlayerImage(playerDiv, playerData.mc || {});

        // Try to extract embed code from pluginData
        const embedCode = playerData.pluginData?.["sharing@web"]?.embedCode;
        if (embedCode) {
          const result = await this.buildHeaderFromEmbedCode(embedCode, isAudioOnly, imageUrl);
          if (result) return result;
        }

        // Fallback: construct player from media URL if available
        const result = await this.buildHeaderFromStreams(streams, isAudioOnly, imageUrl);
        if (result) return result;
      } catch (error) {
        logger.debug({ error }, "Failed to parse Tagesschau media player data");
      }
    }

    return null;
  }

  private getMediaPlayers(soup: cheerio.CheerioAPI): cheerio.Cheerio<any> {
    const mediaPlayers = soup('div[data-v-type="MediaPlayer"]').filter((_, el) => {
      return (soup(el).attr("class") || "").toLowerCase().includes("mediaplayer");
    });

    const teaserPlayers = mediaPlayers.filter((_, el) => (soup(el).attr("class") || "").includes("teaser-top"));
    return teaserPlayers.length > 0 ? teaserPlayers : mediaPlayers;
  }

  private parsePlayerData(dataV: string): any {
    const decoded = dataV
      .replace(/&quot;/g, '"')
      .replace(/&#39;/g, "'")
      .replace(/&amp;/g, "&")
      .replace(/&lt;/g, "<")
      .replace(/&gt;/g, ">");
    return JSON.parse(decoded);
  }

  private getPlayerImage(playerDiv: cheerio.Cheerio<any>, mc: any): string | null {
    let imageUrl = this.getPlayerImageFromMetadata(mc);

    if (!imageUrl) {
      imageUrl = this.getPlayerImageFromDOM(playerDiv);
    }

    if (imageUrl) {
      if (imageUrl.startsWith("//")) return "https:" + imageUrl;
      if (imageUrl.startsWith("/")) return "https://www.tagesschau.de" + imageUrl;
    }
    return imageUrl;
  }

  private getPlayerImageFromMetadata(mc: any): string | null {
    const fields = ["poster", "image", "thumbnail", "preview", "cover"];

    // Check main mc object
    for (const field of fields) {
      if (mc[field]) return mc[field];
    }

    // Check streams
    if (mc.streams) {
      for (const stream of mc.streams) {
        for (const field of fields) {
          if (stream[field]) return stream[field];
        }
      }
    }

    return null;
  }

  private getPlayerImageFromDOM(playerDiv: cheerio.Cheerio<any>): string | null {
    const parent = playerDiv.parent();
    if (parent && parent.length > 0) {
      const img = parent.find("img").first();
      if (img.length > 0) return img.attr("src") || null;
    }

    const prev = playerDiv.prev();
    if (prev && prev.length > 0) {
      const img = prev.find("img").first();
      if (img.length > 0) return img.attr("src") || null;
    }

    return null;
  }

  private async buildHeaderFromEmbedCode(embedCode: string, isAudioOnly: boolean, imageUrl: string | null): Promise<string | null> {
    const decoded = embedCode.replace(/&quot;/g, '"').replace(/&#39;/g, "'").replace(/&amp;/g, "&").replace(/&lt;/g, "<").replace(/&gt;/g, ">");
    const $ = cheerio.load(decoded);
    const iframe = $("iframe").first();
    let src = iframe.attr("src");
    if (!src) return null;

    src = src.replace("$params$", "");
    if (src.startsWith("//")) src = "https:" + src;
    else if (src.startsWith("/")) src = "https://www.tagesschau.de" + src;

    const height = isAudioOnly ? "200" : "315";
    const playerHtml = `<div class="media-player" style="width: 100%;"><iframe src="${src}" width="100%" height="${height}" frameborder="0" allowfullscreen scrolling="no"></iframe></div>`;

    if (isAudioOnly && imageUrl) {
      const imgPart = await this.buildBase64Image(imageUrl);
      return `<header class="media-header">${imgPart}${playerHtml}</header>`;
    }
    return `<header class="media-header">${playerHtml}</header>`;
  }

  private async buildHeaderFromStreams(streams: any[], isAudioOnly: boolean, imageUrl: string | null): Promise<string | null> {
    if (isAudioOnly) {
      const audioMedia = this.findMediaByMimeType(streams, "audio");
      if (audioMedia) {
        const imgPart = imageUrl ? await this.buildBase64Image(imageUrl) : "";
        return `<header class="media-header">${imgPart}<div class="media-player" style="width: 100%;"><audio controls preload="auto" style="width: 100%;"><source src="${audioMedia.url}" type="${audioMedia.mimeType}">Your browser does not support the audio element.</audio></div></header>`;
      }
    } else {
      const videoMedia = this.findMediaByMimeType(streams, "video");
      if (videoMedia) {
        const poster = imageUrl ? `poster="${imageUrl}"` : "";
        return `<header class="media-header"><div class="media-player" style="width: 100%;"><video controls preload="auto" ${poster} style="width: 100%;"><source src="${videoMedia.url}" type="${videoMedia.mimeType}">Your browser does not support the video element.</video></div></header>`;
      }
    }
    return null;
  }

  private findMediaByMimeType(streams: any[], type: string): { url: string; mimeType: string } | null {
    for (const stream of streams) {
      for (const media of (stream.media || [])) {
        if (media.url && (media.mimeType || "").toLowerCase().includes(type)) {
          return { url: media.url, mimeType: media.mimeType };
        }
      }
    }
    return null;
  }

  private async buildBase64Image(imageUrl: string): Promise<string> {
    try {
      const headerElement = await createHeaderElementFromUrl(imageUrl, "Article image");
      if (headerElement) {
        const srcMatch = /src=["']([^"']+)["']/.exec(headerElement);
        if (srcMatch) {
          return `<div class="media-image"><img src="${srcMatch[1]}" alt="Article image" style="max-width: 100%; height: auto; border-radius: 8px;"></div>`;
        }
      }
    } catch { /* fallback */ }
    return `<div class="media-image"><img src="${imageUrl}" alt="Article image" style="max-width: 100%; height: auto; border-radius: 8px;"></div>`;
  }

  protected override async fetchArticleContentInternal(url: string, article: RawArticle): Promise<string> {
    const html = await super.fetchArticleContentInternal(url, article);
    this.originalHtmlCache.set(article.url, html);
    return html;
  }

  protected override async processContent(html: string, article: RawArticle): Promise<string> {
    let mediaHeader: string | null = null;
    const originalHtml = this.originalHtmlCache.get(article.url);

    if (originalHtml) {
      try {
        mediaHeader = await this.extractMediaHeader(cheerio.load(originalHtml));
        this.originalHtmlCache.delete(article.url);
      } catch (error) {
        logger.debug({ error, url: article.url }, "Failed to extract media header");
      }
    }

    const $ = cheerio.load(html);
    $("p, div, span").each((_, el) => {
      const $el = $(el);
      if (!$el.text().trim() && !$el.find("img").length) $el.remove();
    });

    let content = sanitizeHtml($.html() || "");
    if (mediaHeader) content = mediaHeader + content;

    const { processContent: processContentUtil } = await import("./base/process");
    return await processContentUtil(content, article, this.feed?.generateTitleImage ?? true, this.feed?.addSourceFooter ?? true);
  }

  protected override async extractContent(html: string, article: RawArticle): Promise<string> {
    const extracted = this.extractContentFromTextabsatz(html);
    return await super.removeElementsBySelectors(extracted, article);
  }

  private extractContentFromTextabsatz(html: string): string {
    try {
      const $ = cheerio.load(html);
      const content = $('<div class="article-content"></div>');

      $("p, h2").each((_, element) => {
        const $el = $(element);
        if (this.shouldSkipElement($el)) return;

        if (element.tagName === "p" && ($el.attr("class") || "").includes("textabsatz")) {
          content.append($el.clone().removeAttr("class"));
        } else if (element.tagName === "h2" && ($el.attr("class") || "").includes("trenner")) {
          content.append($("<h2></h2>").text($el.text().trim()));
        }
      });

      return content.html() || "";
    } catch (error) {
      logger.error({ error }, "Extraction failed");
      return html;
    }
  }

  private shouldSkipElement($el: cheerio.Cheerio<any>): boolean {
    const parent = $el.parent();
    if (parent.length === 0) return false;
    const classes = parent.attr("class") || "";
    return ["teaser", "bigfive", "accordion", "related"].some(c => classes.includes(c));
  }
}
